src/lcd.c

```diff
@@ -0,0 +1,146 @@
+#include "stm32f0xx_gpio.h"
+#include "stm32f0xx_rcc.h"
+#include "stm32f0xx_spi.h"
+#include "cmsis/cmsis_device.h"
+#include "diag/Trace.h"
+#include "lcd.h"
+
+#define LCD_ENABLE (0x80)
+#define LCD_DISABLE (0x0)
+#define LCD_COMMAND (0x0)
+#define LCD_DATA (0x40)
+
+void LCD_Init(void){
+	// Configure GPIOB to control LCD via SPI
+	myGPIOB_Init();
+	mySPI_Init();
+
+	// Configure the internal LCD controls
+	// PBMCUSLK has a shift register connected to the LCD like:
+	//    Q7  Q6  Q5  Q4  |  Q3  Q2  Q1  Q0
+	//    =================================
+	//    EN  R/S NC  NC  |  D7  D6  D5  D4
+	// When the LCD is initialized it's working in 8bit word mode.
+	// The PBMCUSLK will interpret the 8bit word sent by SPI according to
+	// the above diagram and D3->D0 = 0. The first thing we have to do is
+	// shift into 4b mode operation so we can ignore D3->D0.
+	// We can't use our regular write command / data functions since they're
+	// written with the assumption we're in 4b mode.
+	uint8_t set4bMode = 0x2;
+	HC595_Write(LCD_DISABLE | LCD_COMMAND | set4bMode);
+	HC595_Write(LCD_ENABLE | LCD_COMMAND | set4bMode);
+	HC595_Write(LCD_DISABLE | LCD_COMMAND | set4bMode);
+
+	// Now do the rest of the LCD config
+	// 4 bits, 2 lines, 5x7 font
+	LCD_SendWord(LCD_COMMAND, 0x28);
+
+	// Display ON, No cursors
+	LCD_SendWord(LCD_COMMAND, 0x0E);
+
+	// Entry mode: Increment, No Display shifting
+	LCD_SendWord(LCD_COMMAND, 0x06);
+
+	// Clear display
+	LCD_SendWord(LCD_COMMAND, 0x01);
+
+	// Write the initial units and resistance / freq placeholders manually instead
+	// of trying to pass non-int values through the number printing functions
+	//  ?.?    Ω
+	//  ?.?    H
+	// TODO: Ω and H are static, so draw them in place
+}
+
+void myGPIOB_Init(void){
+	// Turn on the GPIOB clock
+	RCC_AHBPeriphClockCmd(RCC_AHBPeriph_GPIOB, ENABLE);
+
+	// PB3 --AF0-> SPI MOSI
+	// PB5 --AF0-> SPI SCK
+	GPIO_InitTypeDef GPIO_InitStruct;
+	GPIO_InitStruct.GPIO_Pin   = GPIO_Pin_3 | GPIO_Pin_5;
+	GPIO_InitStruct.GPIO_Mode  = GPIO_Mode_AF;
+	GPIO_InitStruct.GPIO_Speed = GPIO_Speed_50MHz;
+	GPIO_InitStruct.GPIO_OType = GPIO_OType_PP;
+	GPIO_InitStruct.GPIO_PuPd  = GPIO_PuPd_NOPULL;
+	GPIO_Init(GPIOB, &GPIO_InitStruct);
+
+	// PB4 will be manually toggled to control the output update on the shift register.
+	// This will expose the contents of the register to the LCD controller.
+	GPIO_InitStruct.GPIO_Pin   = GPIO_Pin_4;
+	GPIO_InitStruct.GPIO_Mode  = GPIO_Mode_OUT;
+	GPIO_InitStruct.GPIO_Speed = GPIO_Speed_50MHz;
+	GPIO_InitStruct.GPIO_OType = GPIO_OType_PP;
+	GPIO_InitStruct.GPIO_PuPd  = GPIO_PuPd_NOPULL;
+	GPIO_Init(GPIOB, &GPIO_InitStruct);
+}
+
+void mySPI_Init(void){
+	// Turn on the SPI clock
+	RCC_APB2PeriphClockCmd(RCC_APB2Periph_SPI1 , ENABLE);
+
+	SPI_InitTypeDef SPI_InitStruct;
+	SPI_InitStruct.SPI_Direction 		 = SPI_Direction_1Line_Tx;
+	SPI_InitStruct.SPI_Mode 			 = SPI_Mode_Master;
```

Comment:
This lines up in Eclipse but, obviously, it's a mix of tabs and spaces. This should be cleaned up 😕